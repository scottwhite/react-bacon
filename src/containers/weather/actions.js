const actions = {
  fetchWeather(latlng){
    return {
      type: 'WEATHER_GET_REQUESTED',
      latlng
    }
  }
}


export default actions
