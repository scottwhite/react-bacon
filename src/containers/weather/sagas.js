import { call, put } from 'redux-saga/effects'
import Api from '../../api'

// worker Saga: will be fired on WEATHER_GET_REQUESTED actions
export function* fetchWeather(action) {
  try {
    const data = yield call(Api.weather, action.latlng)
    yield put({type: 'WEATHER_GET_SUCCEEDED', weather: data})
  } catch (e) {
    yield put({type: 'WEATHER_GET_FAILED', message: e.message})
  }
}
