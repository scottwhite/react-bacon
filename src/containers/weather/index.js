import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import actions from './actions'
import WeatherList from './weather-list'

export class Weather extends Component {
  componentDidMount(){
    let fetchWeather = this.props.fetchWeather
    let latlng = this.props.params.latlng
    fetchWeather(latlng)
  }
  render() {
    const {error, weather} = this.props;
    let err = '';
    if(error){
      err = (<div className='alert alert-danger'>{error}</div>)
    }
    return (
      <div className='container-fluid'>
        {err}
        <WeatherList weather={weather} />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  const { weather, error } = state.weather;
  return {
    weather,
    error
  };
};

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators(actions, dispatch);
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Weather);
