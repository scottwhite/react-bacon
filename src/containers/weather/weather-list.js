import React from 'react'
import './weather.css'
import WeatherIcon from './weather-icon'

export const WeatherList = (props) => {
  let weather = props.weather || {};
  let days = []
  let summary = ''
  if(weather.data){
    days = weather.data
  }
  if(weather.summary){
    summary = weather.summary
  }
  function selectWeather(day){
    props.onClick(day)
  }
  function formatWeatherDate(value){
    let d = new Date(value * 1000)
    return d.toDateString()
  }
  return (
    <div className='container-fluid'>
      <h2>Weather</h2>
      <div>{summary}</div>
      <ul className='weather-list'>
        {
          days.map((day, idx) => (
            <li key={day.time} onClick={selectWeather}>
              <div className='weather-day'>{formatWeatherDate(day.time)}</div>
              <WeatherIcon id={day.time + ''} name={day.icon} />
              <div className='weather-summary'>{day.summary}</div>
              <div className='weather-temp'><label>Low</label> {day.temperatureMin}&deg; F</div>
              <div className='weather-temp'><label>High</label> {day.temperatureMax}&deg; F</div>
            </li>
          ))
        }
      </ul>
    </div>
  )
}

WeatherList.propTypes = {
  weather: React.PropTypes.object.isRequired,
  onClick: React.PropTypes.func
}

export default WeatherList
