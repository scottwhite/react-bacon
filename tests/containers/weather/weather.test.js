import React from 'react';
import renderer from 'react/lib/ReactTestRenderer';

import {Weather} from '../../../src/containers/weather/index';
import weatherActions from '../../../src/containers/weather/actions'
import {forecast} from '../../../mock-data/api.weather.mock'

describe('Weather', () => {
  it('shows the forecast', () => {
    let fetchWeatherReal = weatherActions.fetchWeather
    let search = 'washing'
    let fetchWeatherMock = (latlng)=>{
      let message = fetchWeatherReal(latlng)
      expect(message.type).toBe('WEATHER_GET_REQUESTED')
      expect(message.latlng).toBe(latlng)
    }
    let latlng = '38.97844529999999, -76.4921829'
    const component = renderer.create(
      <Weather fetchWeather={fetchWeatherMock} params={{latlng}} weather={forecast.daily}/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot()
    // BRITTLE!!!
    expect(tree.children[1].children[2].children.length).toBe(forecast.daily.data.length)
  });

  it('shows an error if error', ()=>{
    let error = 'ERROR!'
    let latlng = '38.97844529999999, -76.4921829'
    const component = renderer.create(
      <Weather fetchWeather={()=>{}} params={{latlng}} weather={{}} error={error}/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot()
    expect(tree.children[0].children[0]).toBe(error)
  })

});
