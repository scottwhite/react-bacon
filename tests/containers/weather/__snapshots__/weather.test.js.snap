exports[`Weather shows an error if error 1`] = `
<div
  className="container-fluid">
  <div
    className="alert alert-danger">
    ERROR!
  </div>
  <div
    className="container-fluid">
    <h2>
      Weather
    </h2>
    <div>
      
    </div>
    <ul
      className="weather-list" />
  </div>
</div>
`;

exports[`Weather shows the forecast 1`] = `
<div
  className="container-fluid">
  
  <div
    className="container-fluid">
    <h2>
      Weather
    </h2>
    <div>
      Light rain throughout the week, with temperatures peaking at 93°F on Sunday.
    </div>
    <ul
      className="weather-list">
      <li
        onClick={[Function selectWeather]}>
        <div
          className="weather-day">
          Wed Aug 10 2016
        </div>
        <canvas
          height={128}
          id="1470801600"
          width={128}>
          rain
        </canvas>
        <div
          className="weather-summary">
          Rain in the morning and evening.
        </div>
        <div
          className="weather-temp">
          <label>
            Low
          </label>
           
          77.49
          ° F
        </div>
        <div
          className="weather-temp">
          <label>
            High
          </label>
           
          89.64
          ° F
        </div>
      </li>
      <li
        onClick={[Function selectWeather]}>
        <div
          className="weather-day">
          Thu Aug 11 2016
        </div>
        <canvas
          height={128}
          id="1470888000"
          width={128}>
          rain
        </canvas>
        <div
          className="weather-summary">
          Light rain until afternoon, starting again in the evening.
        </div>
        <div
          className="weather-temp">
          <label>
            Low
          </label>
           
          76.67
          ° F
        </div>
        <div
          className="weather-temp">
          <label>
            High
          </label>
           
          89.78
          ° F
        </div>
      </li>
      <li
        onClick={[Function selectWeather]}>
        <div
          className="weather-day">
          Fri Aug 12 2016
        </div>
        <canvas
          height={128}
          id="1470974400"
          width={128}>
          partly-cloudy-day
        </canvas>
        <div
          className="weather-summary">
          Partly cloudy until afternoon.
        </div>
        <div
          className="weather-temp">
          <label>
            Low
          </label>
           
          77.69
          ° F
        </div>
        <div
          className="weather-temp">
          <label>
            High
          </label>
           
          90.45
          ° F
        </div>
      </li>
      <li
        onClick={[Function selectWeather]}>
        <div
          className="weather-day">
          Sat Aug 13 2016
        </div>
        <canvas
          height={128}
          id="1471060800"
          width={128}>
          rain
        </canvas>
        <div
          className="weather-summary">
          Drizzle in the afternoon.
        </div>
        <div
          className="weather-temp">
          <label>
            Low
          </label>
           
          78.66
          ° F
        </div>
        <div
          className="weather-temp">
          <label>
            High
          </label>
           
          92.76
          ° F
        </div>
      </li>
      <li
        onClick={[Function selectWeather]}>
        <div
          className="weather-day">
          Sun Aug 14 2016
        </div>
        <canvas
          height={128}
          id="1471147200"
          width={128}>
          rain
        </canvas>
        <div
          className="weather-summary">
          Light rain in the morning and overnight.
        </div>
        <div
          className="weather-temp">
          <label>
            Low
          </label>
           
          81.37
          ° F
        </div>
        <div
          className="weather-temp">
          <label>
            High
          </label>
           
          93.15
          ° F
        </div>
      </li>
      <li
        onClick={[Function selectWeather]}>
        <div
          className="weather-day">
          Mon Aug 15 2016
        </div>
        <canvas
          height={128}
          id="1471233600"
          width={128}>
          partly-cloudy-day
        </canvas>
        <div
          className="weather-summary">
          Partly cloudy throughout the day.
        </div>
        <div
          className="weather-temp">
          <label>
            Low
          </label>
           
          79.43
          ° F
        </div>
        <div
          className="weather-temp">
          <label>
            High
          </label>
           
          86.69
          ° F
        </div>
      </li>
      <li
        onClick={[Function selectWeather]}>
        <div
          className="weather-day">
          Tue Aug 16 2016
        </div>
        <canvas
          height={128}
          id="1471320000"
          width={128}>
          rain
        </canvas>
        <div
          className="weather-summary">
          Rain starting in the afternoon.
        </div>
        <div
          className="weather-temp">
          <label>
            Low
          </label>
           
          77.84
          ° F
        </div>
        <div
          className="weather-temp">
          <label>
            High
          </label>
           
          87.91
          ° F
        </div>
      </li>
      <li
        onClick={[Function selectWeather]}>
        <div
          className="weather-day">
          Wed Aug 17 2016
        </div>
        <canvas
          height={128}
          id="1471406400"
          width={128}>
          rain
        </canvas>
        <div
          className="weather-summary">
          Light rain until afternoon, starting again in the evening.
        </div>
        <div
          className="weather-temp">
          <label>
            Low
          </label>
           
          79.01
          ° F
        </div>
        <div
          className="weather-temp">
          <label>
            High
          </label>
           
          86.35
          ° F
        </div>
      </li>
    </ul>
  </div>
</div>
`;
